add ability to translate the user guide on Crowdin

michaelDCurran opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests.

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review".
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
None.

### Summary of the issue:
Currently user documentation such as the user guide are made translatable via a custom (and very old) translation system hosted by NV Access. For many reasons we need to move away from this old system to something more mainstream and  maintainable. We have already successfully moved translation of NVDA interface messages to Crowdin, and we should do the same for the user guide and other documentation.
 
### Description of user facing changes
None.

### Description of development approach
* Added markdownTranslate.py, which contains several commands for generating and updating  xliff files from markdown files. These xliff files can then be uploaded to Crowdin for translation, and eventually downloaded again and converted back to markdown files. Commands include:
    * generateXliff: to generate an xliff file from a markdown file. Firstly a 'skeleton' of the markdown file is produced which is all the structure of a markdown file, but the translatable content on each line has been replaced by a special translation ID. Lines such as blank lines, hidden header rows, or table header separator lines are included in the skeleton in tact and are not available for translation. The xliff file is then produced, which contains one translatable string per translation unit, keyed by its respective translation ID. Each unit also contains translator notes to aide in translation, such as the line number, and any prefix or suffix markdown structure. E.g. a heading might have a prefix of `###` and a suffix of `{#Intro}`. The skeleton is also embedded into the xliff file so that it is possible to update the xliff file keeping existing translation IDs, and or generate the existing markdown file from the xliff file.
    * generateMarkdown: Given an xliff file, a the original markdown file is reproduced from the embedded skeleton, and either the translated or source strings from the xliff file, depending on whether you want a translated or untranslated markdown file.
    * updateXliff: to update an existing xliff file with changes from a markdown file, ensuring that IDs of existing translatable strings are kept in tact. This command extracts the skeleton from the xliff file, makes a diff of the old and new markdown files, then applies this diff to the skeleton file I.e. removes skeleton lines that were removed from the markdown file, and adds skeleton lines (with new IDs) for lines that are newly added  to the markdown file. All existing lines stay as is keeping their existing translation IDs. Finally a new xliff file is generated from the up to date markdown file and skeleton, resulting in an xliff file that contains all translatable strings from the new markdown file, but reusing translation IDs for existing strings. 
    * translateXliff: given an xliff file, and a pretranslated markdown file that matches the skeleton, a new xliff file is produced containing translations for all strings.
    * pretranslateAllPossibleLangs: this walks the NVDA user_docs directory, and for each language, pretranslates the English xliff file using the existing pretranslated markdown file from the old translation system (if it matches the skeleton exactly) producing a translated xliff file that can be uploaded to Crowdin to bring an existing translation up to where it was in the old  system.
    * runTests: runs some tests using 2024.2 and 2024.3 English and French user guides to ensure files are translated and generated as expected.
* Added a generated xliff file for the current English user guide markdown file. Note that this has been uploaded to Crowdin for translation. 
* Added a GitHub action that runs on the beta branch if English userGuide.md changes. The action regenerates the original markdown file from the current English user guide xliff, then updates the xliff file based on the changes from the original markdown file to the current markdown file. This xliff file is then uploaded to Crowdin, and also committed and pushed to beta.
 
### Testing strategy:
* Uploaded the English user guide xliff to Crowdin.
* pretranslated a French xliff file from the English user guide xliff and a pretranslated French markdown file. 
* Uploaded the pretranslated French  xliff file to Crowdin for French.
* Downloaded the latest French xliff file from Crwdin, ensuring all strings were translated.
* Added / removed several lines from the English user guide markdown file and committed changes.
* Ensured the GitHub action ran with no errors.
* Downloded the French xliff from Crowdin and confirmed that the new added strings were available for translation, the removed strings were no longer available, and that all existing strings were still translated appropriately.

### Known issues with pull request:
None known.

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced an automated workflow for updating user documentation based on changes in markdown files, enhancing documentation consistency.
  
- **Bug Fixes**
  - Improved formatting and readability of various functions and classes across the codebase without altering functionality.
  
- **Documentation**
  - Enhanced clarity in comments and docstrings throughout the code, contributing to better maintainability. 

- **Chores**
  - General code style improvements, including consistent spacing and formatting adjustments to adhere to coding standards.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

michaelDCurran committed:
Add markdownTranslate script which can generate skel and xlif files from markdown files.

michaelDCurran committed:
markdownTranslate.py: remove unneeded code.

michaelDCurran committed:
markdownTranslation: allow bad heading anchors in translations.

michaelDCurran committed:
markdownTranslation: ignore differences in cell padding when comparing post table header line and hidden header row.

michaelDCurran committed:
sconstruct: ruild localized md files in user_docs from a localized xliff and English skel files.

michaelDCurran committed:
Add github action to regenerate English user documentation xliff and skel files if English user documentation markdown files change.

michaelDCurran committed:
markdownTranslate: when generating an xliff file, embed the skeleton file so that it is theoretically possible to regenerate the markdown withonly the xliff. generateMarkdown and translateXliff commands now read the skeleton straight from the xliff file, rather than receiving it as an argument.

michaelDCurran committed:
regenerate english documentation xliff: update for changed markdownTranslate commands.

michaelDCurran committed:
github action: add comments. Paass Github org/repo to generateXliff command.

michaelDCurran committed:
markdownTranslate.generateXliff: the skeleton must appear as the first child of the file tag.

michaelDCurran committed:
Add test files for markdownTranslate

michaelDCurran committed:
markdownTranslate generateMarkdown no longer takes a skel file argument, as it already reads the skeleton content from the xliff.

michaelDCurran committed:
user docs github action: upload to crowdin.

michaelDCurran committed:
Add generated English userguide.xliff

michaelDCurran committed:
test update to english user guide

michaelDCurran committed:
User docs github action: correct formatting.

michaelDCurran committed:
test update english user guide

michaelDCurran committed:
Fix comments

michaelDCurran committed:
test update english user guide

michaelDCurran committed:
Refactoring of markdownTranslate.py. Much simpler commands.

michaelDCurran committed:
Update user docs github action for new markdownTranslate updateXliff command

michaelDCurran committed:
Test update english user guide

michaelDCurran committed:
Fix variable syntax

michaelDCurran committed:
Test update english user guide

michaelDCurran committed:
Fix variable syntax

michaelDCurran committed:
Test update english user guide

michaelDCurran committed:
user docs github action:  requests is needed by crowdinSync.py

michaelDCurran committed:
Test update english user guide

michaelDCurran committed:
Fix user docs github action.

michaelDCurran committed:
Test update english user guide

actions-user committed:
Update user documentation files

michaelDCurran committed:
user docs github action: new Crowdin file ID

michaelDCurran committed:
Test update english user guide

michaelDCurran committed:
user docs github action: new Crowdin file ID

michaelDCurran committed:
Test update english user guide

actions-user committed:
Update user documentation files

michaelDCurran committed:
markdownTranslate: add final new line for embedded skeleton.

michaelDCurran committed:
Test update english user guide

actions-user committed:
Update user documentation files

michaelDCurran committed:
English user guide: restore to beta

actions-user committed:
Update user documentation files

michaelDCurran committed:
Test update english user guide

actions-user committed:
Update user documentation files

michaelDCurran committed:
user docs github action: limit to just the beta branch

michaelDCurran committed:
Restore english user guide bck to beta

michaelDCurran committed:
user docs github action: allow running on xliff branch as well for no until merged.

michaelDCurran committed:
Add test string again.

actions-user committed:
Update user documentation files

michaelDCurran committed:
Restore english user guide back to beta

actions-user committed:
Update user documentation files

michaelDCurran committed:
Test string added to english user guide.

actions-user committed:
Update user documentation files

michaelDCurran committed:
Restore english user guide back to beta

actions-user committed:
Update user documentation files

michaelDCurran commented:
@seanbudd  It seems that precommit is reformatting all files in the repo which makes this pr impossible to review. I guess I'll need to wait for all the reformatting to reach beta and then remove the final precommit here before I can make it ready for review?

CyrilleB79 commented:
Before merging this PR, could you send a message to the translators mailing list and give a deadline for them to provide up-to-date translations?
You should explain that it's the last chance for them to reach the identical structure state, allowing to benefit from automatic translation in the new file.
Thanks.

dpy013 commented:
> Before merging this PR, could you send a message to the translators mailing list and give a deadline for them to provide up-to-date translations? You should explain that it's the last chance for them to reach the identical structure state, allowing to benefit from automatic translation in the new file. Thanks.
Suggest updating as well:
#15014 
Update on the plan

michaelDCurran committed:
English user guide: add test string.

actions-user committed:
Update user documentation files

michaelDCurran commented:
@CyrilleB79  and @dpy013   I'll definitely be providing a clear and detailed update to the translators list and the discussion once this PR is finally finalized approved and merged. But until then, things could change so I do not want to say anything prematurely. Please be assured though that we'll give translators plenty of time. Plus, the system is such that we can move the languages to Crowdin that are ready -- currently about 17. And when more become complete, we can move those as well.

michaelDCurran committed:
Restore english user guide to beta.

michaelDCurran committed:
User docs github action: for now only handle user guide, not changes (what's new).

michaelDCurran committed:
English user guide: add test string.

actions-user committed:
Update user documentation files

michaelDCurran committed:
Restore english user guide to beta

actions-user committed:
Update user documentation files

michaelDCurran committed:
Merge branch 'beta' into xliff

seanbudd commented on user_docs/markdownTranslate.py:
```diff
@@ -0,0 +1,734 @@
+# A part of NonVisual Desktop Access (NVDA)
```
I still need to review this file further

seanbudd commented on user_docs/markdownTranslate.py:
```diff
@@ -0,0 +1,734 @@
+		help="The base name of the markdown files to pretranslate",
+	)
+	args = mainParser.parse_args()
+	if args.command == "generateXliff":
```
this could probably use match/case syntax

seanbudd commented on user_docs/markdownTranslate.py:
```diff
@@ -0,0 +1,734 @@
+	subprocess.run(["python", __file__, command, *args], check=True)
+
+
+def runTests(testDir: str):
```
could these tests be moved to our standard unit tests?

michaelDCurran committed:
user docs github action: add comment.

michaelDCurran committed:
Convert markdownTranslate tests to unit tests.

michaelDCurran committed:
markdownTranslate unit tests: refactor so as they stop on the first error.

pre-commit-ci[bot] committed:
Pre-commit auto-fix

michaelDCurran committed:
User docs github action: refactoring to ensure that:
1. The upload to Crowdin only happens after a successful  git commit / push, as opposed to the other way around. Otherwise if the push fails, subsequent runs may create new translation IDs for already translated strings, causing needless work for translators.
 2. Check the entire commit range of the push that caused the action to see if a markdown file has changed. Previously only the final commit was being checked.

 Also add more comments and logging.

michaelDCurran committed:
English user guide: add test string

actions-user committed:
Update user documentation files

michaelDCurran committed:
Restore english user guide to beta

michaelDCurran committed:
English changes.md: add test string

actions-user committed:
Update user documentation files

michaelDCurran committed:
Restore English changes file back to beta

michaelDCurran committed:
user docs github action: Don't fail the job if no xliff files were changed.

michaelDCurran committed:
Test changes.md

michaelDCurran committed:
Restore English changes back to beta

michaelDCurran committed:
English user guide test string

actions-user committed:
Update user_docs/en/userGuide.xliff

michaelDCurran committed:
Restore english user guide back to beta

actions-user committed:
Update user_docs/en/userGuide.xliff

michaelDCurran committed:
user docs github action: add more logging

seanbudd commented on user_docs/markdownTranslate.py:
```diff
@@ -0,0 +1,670 @@
+import subprocess
+
+RAW_GITHUB_REPO_URL = "https://raw.githubusercontent.com/nvaccess/nvda"
+re_kcTitle = re.compile(r"^(<!--\s+KC:title:\s*)(.+?)(\s*-->)$")
+re_kcSettingsSection = re.compile(r"^(<!--\s+KC:settingsSection:\s*)(.+?)(\s*-->)$")
+re_comment = re.compile(r"^<!--.+-->$")
+re_heading = re.compile(r"^(#+\s+)(.+?)((?:\s+\{#.+\})?)$")
+re_bullet = re.compile(r"^(\s*\*\s+)(.+)$")
+re_number = re.compile(r"^(\s*[0-9]+\.\s+)(.+)$")
+re_hiddenHeaderRow = re.compile(r"^\|\s*\.\s*\{\.hideHeaderRow\}\s*(\|\s*\.\s*)*\|$")
+re_postTableHeaderLine = re.compile(r"^(\|\s*-+\s*)+\|$")
+re_tableRow = re.compile(r"^(\|)(.+)(\|)$")
+re_translationID = re.compile(r"^(.*)\$\(ID:([0-9a-f-]+)\)(.*)$")
```
I think these should be grouped as an enum

```py
class Re(re.Pattern, Enum):
   KC_TITLE  = re.compile(r"^(<!--\s+KC:title:\s*)(.+?)(\s*-->)$")

seanbudd replied:
@michaelDCurran this is the only remaining feedback from me

michaelDCurran replied:
A few issues when trying to do this:
* At runtime:
```
class MarkdownRegex(re.Pattern, enum.Enum):                                                                                                             
  File "C:\Users\mick\AppData\Local\Programs\Python\Python311-32\Lib\enum.py", line 562, in __new__                                                         
    enum_class = super().__new__(metacls, cls, bases, classdict, **kwds)                                                                                    
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                    
TypeError: type 're.Pattern' is not an acceptable base type 
```
* In VS code, Pylance says:
```
Error: Base class "Pattern[Unknown]" is marked final and cannot be subclassed at line 25 and character 21. generated by Pylance
```
* If I don't inherit from re.Pattern however, then I of course get errors in VS Code like:
```
Error: Cannot access attribute "match" for class "Literal[MarkdownRegex.HIDDEN_HEADER_ROW]"   Attribute "match" is unknown at line 93 and character 38. generated by Pylance
```
So based on this, for now perhaps we should leave them as stand-alone constants?

seanbudd replied:
Makes sens

seanbudd commented on user_docs/markdownTranslate.py:
```diff
@@ -0,0 +1,670 @@
+	return gitDir
+
+
+def getRawGithubURLForPath(filePath: str):
```
```suggestion
def getRawGithubURLForPath(filePath: str) -> str:
```

seanbudd commented on user_docs/markdownTranslate.py:
```diff
@@ -0,0 +1,670 @@
+	numTranslatedStrings = 0
+
+
+def generateMarkdown(xliffPath: str, outputPath: str, translated: bool = True):
```
```suggestion
def generateMarkdown(xliffPath: str, outputPath: str, translated: bool = True) -> Result_generateMarkdown:
```

seanbudd commented on user_docs/markdownTranslate.py:
```diff
@@ -0,0 +1,670 @@
+					res.numRemovedTranslationPlaceholders += 1
+			else:
+				raise ValueError(f"Unexpected diff line: {mdDiffLine}")
+		print(
+			f"Updated skeleton file with {res.numAddedLines} added lines ({res.numAddedTranslationPlaceholders} translation placeholders), {res.numRemovedLines} removed lines ({res.numRemovedTranslationPlaceholders} translation placeholders), and {res.numUnchangedLines} unchanged lines ({res.numUnchangedTranslationPlaceholders} translation placeholders)"
+		)
```
```suggestion
		print(
			f"Updated skeleton file with {res.numAddedLines} added lines "
			f"({res.numAddedTranslationPlaceholders} translation placeholders), "
			f"{res.numRemovedLines} removed lines ({res.numRemovedTranslationPlaceholders} translation placeholders), "
			f"and {res.numUnchangedLines} unchanged lines ({res.numUnchangedTranslationPlaceholders} translation placeholders)"
		)
```

seanbudd committed:
Apply suggestions from code review

seanbudd commented on user_docs/markdownTranslate.py:
```diff
@@ -0,0 +1,673 @@
+			os.remove(langXliffPath)
+			continue
+		try:
+			ensureFilesMatch(rebuiltLangMdPath, langPretranslatedMdPath, allowBadAnchors=True)
```
```suggestion
			ensureMarkdownFilesMatch(rebuiltLangMdPath, langPretranslatedMdPath, allowBadAnchors=True)
```

michaelDCurran committed:
update user_docs/markdownTranslate.py

Co-authored-by: Sean Budd <sean@nvaccess.org>

michaelDCurran committed:
Merge branch 'beta' into xliff

actions-user committed:
Update user_docs/en/userGuide.xliff

michaelDCurran committed:
Update .github/workflows/regenerate_english_userDocs_translation_source.yml

Co-authored-by: Sean Budd <sean@nvaccess.org>

michaelDCurran committed:
linting

michaelDCurran commented:
@coderabbitai resolve
